### PR TITLE
resolver/manual: support restarts, required for channel idleness

### PR DESCRIPTION
**Pull request by easwars**
RELEASE NOTES: none

### CODE REVIEW

File: resolver/manual/manual.go [hunk 1]
```diff
@@ -58,30 +58,34 @@ type Resolver struct {
...
+	r.BuildCallback(target, cc, opts)
 	r.mu.Lock()
 	r.CC = cc
-	r.mu.Unlock()
-	r.BuildCallback(target, cc, opts)
-	if r.bootstrapState != nil {
-		r.UpdateState(*r.bootstrapState)
+	if r.lastSeenState != nil {
+		err := r.CC.UpdateState(*r.lastSeenState)
+		r.UpdateStateCallback(err)
```
Comment (easwars) [inline resolver/manual/manual.go:82]: I'm invoking this callback with the lock held. I don't see a problem with it. Let me know if you do. Or we can be safe and spawn a goroutine to invoke the callback.

Comment (easwars) [reply resolver/manual/manual.go]: Went ahead and did that anyways.

File: resolver/manual/manual.go [hunk 2]
```diff
@@ -58,30 +58,34 @@ type Resolver struct {
...
+	r.BuildCallback(target, cc, opts)
 	r.mu.Lock()
 	r.CC = cc
-	r.mu.Unlock()
-	r.BuildCallback(target, cc, opts)
-	if r.bootstrapState != nil {
-		r.UpdateState(*r.bootstrapState)
+	if r.lastSeenState != nil {
+		err := r.CC.UpdateState(*r.lastSeenState)
+		go func(err error) { r.UpdateStateCallback(err) }(err)
```
Comment (dfawley) [inline resolver/manual/manual.go:82]: What's the concern about using `go r.UpdateStateCallback(err)`?  `err` can't be modified after this.

File: resolver/manual/manual.go [hunk 3]
```diff
@@ -26,7 +26,7 @@ import (
 	"google.golang.org/grpc/resolver"
 )
 
-// NewBuilderWithScheme creates a new test resolver builder with the given scheme.
+// NewBuilderWithScheme creates a new manual resolver builder with the given scheme.
```
Comment (dfawley) [inline resolver/manual/manual.go:29]: Should we document that this may only be used with a single `ClientConn`?  Otherwise bad things happen.

Comment (easwars) [reply resolver/manual/manual.go:29]: Done.

File: resolver/manual/manual.go (hunk 2 above)
Comment (easwars) [reply resolver/manual/manual.go:82]: I think I was paranoid beyond repair :P

File: resolver/manual/manual.go [hunk 4]
```diff
@@ -26,7 +26,9 @@ import (
 	"google.golang.org/grpc/resolver"
 )
 
-// NewBuilderWithScheme creates a new manual resolver builder with the given scheme.
+// NewBuilderWithScheme creates a new manual resolver builder with the given
+// scheme. Every instance of the manual resolver may only ever be used with a
+// single grpc.ClientConn. Otherwise, bad things will happen.
```
Comment (dfawley) [inline resolver/manual/manual.go:31]: LMAO.

**Review by dfawley — APPROVED** [review]

**Pull request merged**